### PR TITLE
Layout

### PR DESCRIPTION
**Pull request by IZablah**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,23 +1,28 @@
-import logo from './logo.svg';
+import { Route, Switch, Redirect } from 'react-router-dom';
+
```
Comment (TaimaTask) [inline src/App.js]: formate the code

File: src/components/layout/mainNavbar.jsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import logo from '../../images/educationSchoolLogo.jpg';
+import classes from './mainNavbar.module.css';
+
```
Comment (TaimaTask) [inline src/components/layout/mainNavbar.jsx]: format

File: src/components/layout/navbar.jsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import MainNavbar from "./mainNavbar";
+import { Fragment } from "react";
+
+
+const Navbar = (props) => {
```
Comment (TaimaTask) [inline src/components/layout/navbar.jsx]: format

File: src/index.js [hunk 4]
```diff
@@ -3,10 +3,13 @@ import ReactDOM from 'react-dom';
 import './index.css';
 import App from './App';
 import reportWebVitals from './reportWebVitals';
+import { BrowserRouter } from 'react-router-dom';
 
 ReactDOM.render(
   <React.StrictMode>
```
Comment (TaimaTask) [inline src/index.js]: format

File: src/pages/home/homePageForm.module.css [hunk 5]
```diff
@@ -0,0 +1,8 @@
+.image {
+    display: block;
+    width: 50%;
+    border-radius: 50%;
+    padding-top: 1.5rem;
```
Comment (TaimaTask) [inline src/pages/home/homePageForm.module.css]: why rem?

File: src/App.js [hunk 6]
```diff
@@ -1,23 +1,30 @@
-import logo from './logo.svg';
+import { Route, Switch, Redirect } from 'react-router-dom';
+import Navbar from './components/layout/navbar';
+import HomePageForm from './pages/home/homePageForm';
+import SignupPageForm from './pages/signUp/signupPageForm';
+
```
Comment (bqaddumi) [inline src/App.js]: Remove

File: src/App.js [hunk 7]
```diff
@@ -1,23 +1,30 @@
...
-        </a>
-      </header>
-    </div>
+    <Navbar>
+      <Switch>
+        <Route path='/home' exact>
+          <HomePageForm />
+        </Route>
+        <Route path='/login' exact>
+          {/* <HomePageForm /> */}
```
Comment (bqaddumi) [inline src/App.js]: remove comment

File: src/App.js [hunk 8]
```diff
@@ -1,23 +1,30 @@
...
+        </Route>
+        <Route path='/signup' exact>
+          <SignupPageForm />
+        </Route>
+        <Route path='*' exact>
+          <Redirect to='/home' />
+        </Route>
```
Comment (bqaddumi) [inline src/App.js]: only one route should be exact

File: src/components/authentication/configureFirebase.jsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+const AuthContext = React.createContext(
+    {
+        token: '',
+        isLoggedIn: true,
+        login: (token) => { },
+        logout: () => { },
+    }
+);
+
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: We want to use redux not context.

File: src/components/authentication/configureFirebase.jsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+export const AuthContextProvider = (props) => {
+    let initialToken;
+
+    const [token, setToken] = useState(initialToken);
+
+    const userIsLoggedIn = !!token;
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: why `!!`

File: src/components/authentication/configureFirebase.jsx [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    const logoutHandler = useCallback(() => {
+        setToken(null);
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: it should be `''` not `null`

File: src/components/authentication/configureFirebase.jsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        localStorage.removeItem('token');
+    }, []);
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: why empty array? `[]`

File: src/components/authentication/configureFirebase.jsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+    const loginHandler = (token) => {
+        setToken(token);
+        localStorage.setItem('token', token);
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: what do you mean by `localStorage`?

File: src/components/authentication/configureFirebase.jsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    };
+
+    const contextValue = {
+        token: token,
+        isLoggedIn: userIsLoggedIn,
+        login: loginHandler,
+        logout: logoutHandler,
+    };
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: use redux

File: src/components/authentication/configureFirebase.jsx [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+    return (
+        <AuthContext.Provider value={contextValue}>
+            {props.children}
+        </AuthContext.Provider>
```
Comment (bqaddumi) [inline src/components/authentication/configureFirebase.jsx]: use redux

File: src/components/layout/mainNavbar.jsx [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+                }
+                {isLoggedIn &&
+                    <li>
+                        <img src={logo} alt="Logo" className={classes.image} />
+                    </li>
+                }
+                {isLoggedIn &&
+                    <li>
+                        <button onClick={logoutHandler}>Logout</button>
+                    </li>
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.jsx]: we can do it in one condition `isLoggedIn && <li>...</li><li>...</li><li>...</li>`

File: src/components/layout/mainNavbar.jsx [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+                {!isLoggedIn &&
+                    <li>
+                        <NavLink to='/login'>Login</NavLink>
+                    </li>
+                }
+                {!isLoggedIn &&
+                    <li>
+                        <NavLink to='/signup'>Signup</NavLink>
+                    </li>
+                }
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.jsx]: we can do it in one condition `isLoggedIn ?<li>...</li><li>...</li><li>...</li> : <li>...</li> <li>...</li> <li>...</li> `

File: src/components/layout/mainNavbar.module.css [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  width: 100%;
+  background-color: rgb(235, 235, 235);
+  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.2);
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+}
+
+.logo {
+  font-family: "Lato", sans-serif;
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css]: from where did you copy this code? do you understand it?

File: src/components/layout/mainNavbar.module.css [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+  font-size: 2rem;
+  padding-left: 50px;
+}
+
+.header ul {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css]: Please remove `ul` and add specific className

File: src/components/layout/mainNavbar.module.css [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  list-style: none;
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css:18]: What is this?

File: src/components/layout/mainNavbar.module.css [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+  display: flex;
+  padding-right: 50px;
+  align-items: center;
+
+}
+
+.header li {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css]: Please remove `li` and add specific className

File: src/components/layout/mainNavbar.module.css [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+  margin: 0 2rem;
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css]: what is `rem`?

File: src/components/layout/mainNavbar.module.css [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+.header a,
+div {
+  text-decoration: none;
+  color: #000000;
+  /* font-weight: bold; */
+}
+
+.header button {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css]: Add className for each `a div button` do this for all general tags

File: src/components/layout/navbar.jsx [hunk 24]
```diff
@@ -0,0 +1,15 @@
+import MainNavbar from "./mainNavbar";
+import { Fragment } from "react";
```
Comment (bqaddumi) [inline src/components/layout/navbar.jsx]: why is this needed?

File: src/index.js [hunk 25]
```diff
@@ -3,11 +3,15 @@ import ReactDOM from 'react-dom';
 import './index.css';
 import App from './App';
 import reportWebVitals from './reportWebVitals';
+import { BrowserRouter } from 'react-router-dom';
+import { AuthContextProvider } from './components/authentication/configureFirebase';
```
Comment (bqaddumi) [inline src/index.js]: use redux

File: src/pages/home/homePageForm.jsx [hunk 26]
```diff
@@ -0,0 +1,13 @@
+import { Fragment } from "react";
```
Comment (bqaddumi) [inline src/pages/home/homePageForm.jsx]: No need for this

File: src/pages/home/homePageForm.jsx [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+import classes from './homePageForm.module.css';
+import logo from '../../images/educationSchoolLogo.jpg';
+
+const homePageForm = () => {
+    return(
+        <Fragment>
```
Comment (bqaddumi) [inline src/pages/home/homePageForm.jsx]: remove

File: src/pages/home/homePageForm.module.css [hunk 28]
```diff
@@ -0,0 +1,8 @@
+.image {
+    display: block;
+    width: 50%;
+    border-radius: 50%;
+    padding-top: 5px;
+    margin-left: auto;
+    margin-right: auto;
```
Comment (bqaddumi) [inline src/pages/home/homePageForm.module.css]: remove, try to not copy the code

File: src/pages/signUp/signupPageForm.jsx [hunk 29]
```diff
@@ -0,0 +1,36 @@
+import classes from './signupPageForm.module.css';
+import { useRef } from 'react';
+import { useContext } from 'react';
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.jsx]: use redux

File: src/pages/signUp/signupPageForm.jsx [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+import { useHistory } from 'react-router-dom';
+import AuthContext from '../../components/authentication/configureFirebase';
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.jsx]: use redux

File: src/pages/signUp/signupPageForm.jsx [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+
+const SignupPageForm = () => {
+    const history = useHistory();
+    const userNameInputRef = useRef();
+    const authCtx = useContext(AuthContext);
+
+
+    const submitHandler = (event) => {
+        event.preventDefault();
+        // const enteredUserName = userNameInputRef.current.value;
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.jsx]: remove

File: src/pages/signUp/signupPageForm.jsx [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+        authCtx.login('test');
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.jsx]: do not use test try to use some name for example `'Bashar'`

File: src/pages/signUp/signupPageForm.module.css [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+  margin-top: 2rem;
+  align-items: center;
+}
+
+.actions button {
+  cursor: pointer;
+  font: inherit;
+  color: black;
+  background-color: rgb(235, 235, 235);
+  border: 1px solid #453e4a;
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.module.css]: you can use one way `hexa` or `rgb` choose one to all the project

**Review by bqaddumi — CHANGES_REQUESTED** [review]

File: package.json [hunk 34]
```diff
@@ -3,11 +3,15 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@reduxjs/toolkit": "^1.6.0",
     "@testing-library/jest-dom": "^5.11.4",
     "@testing-library/react": "^11.1.0",
     "@testing-library/user-event": "^12.1.10",
+    "jquery": "^3.6.0",
```
Comment (bqaddumi) [inline package.json:10]: not good, where do you use this?

File: package.json [hunk 35]
```diff
@@ -3,11 +3,15 @@
...
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
+    "react-redux": "^7.2.4",
+    "react-router-dom": "^5.2.0",
```
Comment (bqaddumi) [inline package.json:14]: good job

File: src/components/layout/mainNavbar.jsx [hunk 36]
```diff
@@ -0,0 +1,36 @@
...
+                </li>
+                <li className={classes.navList}>
+                    <button onClick={logoutHandler}>Logout</button>
+                </li>
+                <li className={classes.navList}>
+                    <NavLink to='/login'>Login</NavLink>
+                </li>
+                <li className={classes.navList}>
+                    <NavLink to='/signup'>Signup</NavLink>
+                </li>
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.jsx:30]: I think you can put them in a const array and map them here like this

```
const navItems = [<div>userName</div>, <img src={logo} alt="Logo" className={classes.image} />, <button onClick={logoutHandler}>Logout</button>, <NavLink to='/login'>Login</NavLink>, ...]

navItems.map(item =>{
return <li>{item}</li>
})
```

File: src/components/layout/mainNavbar.module.css [hunk 37]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+.logo {
+  font-family: sans-serif;
+  font-size: 2rem;
+  padding-left: 10px;
+  text-transform: uppercase;
+}
+
+.navContainerList {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css:17]: navList is enough

File: src/components/layout/mainNavbar.module.css [hunk 38]
```diff
@@ -0,0 +1,74 @@
...
+  font-size: 1.5rem;
+}
+
+.header a{
+  display: flex;
+  text-decoration: none;
+  color: #000000;
+}
+
+.header button {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css:35]: add class name to button and don't use general style for html components

File: src/components/layout/mainNavbar.module.css [hunk 39]
```diff
@@ -0,0 +1,74 @@
...
+  font: inherit;
+  background-color: transparent;
+  color: #000000;
+  padding: 0.5rem 0.5rem;
+  border-radius: 6px;
+  cursor: pointer;
+}
+
+.header a:hover {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css:44]: add class name to `a` and don't use general style for html components

File: src/components/layout/mainNavbar.module.css [hunk 40]
```diff
@@ -0,0 +1,74 @@
...
+  color: #243bbe;
+}
+
+.header button:hover {
```
Comment (bqaddumi) [inline src/components/layout/mainNavbar.module.css:48]: add class name to button and don't use general style for html components

File: src/pages/home/homePage.jsx [hunk 41]
```diff
@@ -0,0 +1,12 @@
+import classes from './homePage.module.css';
+import logo from '../../images/educationSchoolLogo.jpg';
+
+const homePage = () => {
```
Comment (bqaddumi) [inline src/pages/home/homePage.jsx:4]: `HomePage` component name should start with Camel case

File: src/pages/signIn/signinPageForm.jsx [hunk 42]
```diff
@@ -0,0 +1,7 @@
+const signinPageForm = () => {
```
Comment (bqaddumi) [inline src/pages/signIn/signinPageForm.jsx:1]: Component name should start with Camel case

File: src/pages/signUp/signupPageForm.jsx [hunk 43]
```diff
@@ -0,0 +1,30 @@
+import { useRef } from 'react';
+import { useHistory } from 'react-router-dom';
+import classes from './signupPageForm.module.css';
+
+const SignupPageForm = () => {
+    const history = useHistory();
+    const userNameInputRef = useRef();
+
+    const submitHandler = (event) => {
+        event.preventDefault();
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.jsx:10]: What is this for?

File: src/pages/signUp/signupPageForm.module.css [hunk 44]
```diff
@@ -0,0 +1,54 @@
...
+  max-width: 35rem;
+  border-radius: 6px;
+  background-color: #ebebeb;
+  box-shadow: 0 1px 4px #000000;
+  padding: 1.5rem;
+  text-align: center;
+  color: black;
+}
+
+.control label {
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.module.css:12]: please add className to lable

File: src/pages/signUp/signupPageForm.module.css [hunk 45]
```diff
@@ -0,0 +1,54 @@
...
+  display: block;
+  color: black;
+  font-weight: bold;
+  margin-top: 1.5rem;
+  margin-bottom: 0.5rem;
+}
+
+.control input {
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.module.css:20]: add classname to input

File: src/pages/signUp/signupPageForm.module.css [hunk 46]
```diff
@@ -0,0 +1,54 @@
...
+  text-align: center;
+  padding: 0.5rem;
+}
+
+.actions {
+  margin-top: 2rem;
+  align-items: center;
+}
+
+.actions button {
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.module.css:36]: add class name to button

File: src/pages/signUp/signupPageForm.module.css [hunk 47]
```diff
@@ -0,0 +1,54 @@
...
+  cursor: pointer;
+  font: inherit;
+  color: black;
+  background-color: #ebebeb;
+  border: 1px solid #453e4a;
+  border-radius: 4px;
+  padding: 0.5rem 2.5rem;
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.module.css:43]: use static padding not rem `ex: 20px`

File: src/pages/signUp/signupPageForm.module.css [hunk 48]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+.actions button:hover {
```
Comment (bqaddumi) [inline src/pages/signUp/signupPageForm.module.css:46]: add class name to button

**Review by bqaddumi — CHANGES_REQUESTED** [review]

File: src/pages/signUp/signupPageForm.jsx (hunk 43 above)
Comment (IZablah) [reply src/pages/signUp/signupPageForm.jsx:10]: `event.preventDefault()`: The main function, the method tells the user agent that if the event does not get explicitly handled, its default action should not be taken as it normally would be. The event continues to propagate as usual.

File: package.json (hunk 34 above)
Comment (IZablah) [reply package.json:10]: Sure, I'll fix it.

File: src/components/layout/mainNavbar.module.css (hunk 20 above)
Comment (IZablah) [reply src/components/layout/mainNavbar.module.css:18]: Because I want to hide the dot on the link website.

File: src/components/layout/mainNavbar.module.css (hunk 18 above)
Comment (IZablah) [reply src/components/layout/mainNavbar.module.css]: Sure, I did copy that from the logo w3school, but I will fix it.

File: src/components/authentication/configureFirebase.jsx (hunk 10 above)
Comment (IZablah) [reply src/components/authentication/configureFirebase.jsx]: Because I want to check if the token has been stored data or not:
If yes token is stored then return true;
If no token is stored then return false;

File: src/components/authentication/configureFirebase.jsx (hunk 13 above)
Comment (IZablah) [reply src/components/authentication/configureFirebase.jsx]: Is a web storage object to store the data on the user’s computer locally, which means the stored data is saved across browser sessions and the data stored has no expiration time.

// store user details and jwt token in local storage to keep the user logged in between page refreshes.

File: src/components/authentication/configureFirebase.jsx (hunk 12 above)
Comment (IZablah) [reply src/components/authentication/configureFirebase.jsx]: Will return a memoized version of the callback that only changes if one of the dependencies has changed. That is an array of dependencies working, but in our function, I didn't have any changes.
